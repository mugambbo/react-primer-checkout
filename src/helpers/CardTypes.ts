export interface Card {
    name: string;
    pattern: string;
    mask: string;
    cvcMask: string;
}
  
interface CardType {
  DINERS: Card;
  JCB: Card;
  AMEX: Card;
  VISA: Card;
  MASTERCARD: Card;
  DISCOVER: Card;
  HIPERCARD: Card;
  UNIONPAY: Card;
  MAESTROCARD: Card;
  VERVECARD: Card;
  GENERIC: Card;
}
  
export const CardTypes: CardType = {
  DINERS: {
    name: "DINERS",
    pattern: "^3(?:0[0-59]{1}|[689])[0-9]{0,}$",
    mask: "XXXX XXXX XXXX XX",
    cvcMask: "XXX",
  },
  JCB: {
    name: "JCB",
    pattern: "^(?:2131|1800|35)[0-9]{0,}$",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  },
  AMEX: {
    name: "AMEX",
    pattern: "^3[47][0-9]{0,}$",
    mask: "XXXX XXXXXX XXXXX",
    cvcMask: "XXXX",
  },
  VISA: {
    name: "VISA",
    pattern: "^4[0-9]{6,}$",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  },
  MASTERCARD: {
    name: "MASTERCARD",
    pattern: "^(5[1-5]|222[1-9]|22[3-9]|2[3-6]|27[01]|2720)[0-9]{0,}$",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  },  
  DISCOVER: {
    name: "DISCOVER",
    pattern: "^(6011|65|64[4-9]|62212[6-9]|6221[3-9]|622[2-8]|6229[01]|62292[0-5])[0-9]{0,}$",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  },
  HIPERCARD: {
    name: "HIPERCARD",
    pattern: "^(606282|3841)[0-9]{5,}$",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  },
  UNIONPAY: {
    name: "UNIONPAY",
    pattern: "^(62|88)[0-9]{5,}$",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  },
  MAESTROCARD: {
    name: "MAESTROCARD",
    pattern: "^(5018|5020|5038|6304|6759|6761|6763)[0-9]{8,15}$",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  },
  VERVECARD: {
    name: "VERVECARD",
    pattern: "^((506(0|1))|(507(8|9))|(6500))[0-9]{12,15}$",
    mask: "XXXX XXXX XXXX XXXX XXX",
    cvcMask: "XXX",
  },
  GENERIC: {
    name: "GENERIC",
    pattern: "",
    mask: "XXXX XXXX XXXX XXXX",
    cvcMask: "XXX",
  }
}