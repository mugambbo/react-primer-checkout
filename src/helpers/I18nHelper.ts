/**
 * Checks the browser navigator for the browser language
 * @returns The language code (e.g. en) of the browser
 */
export const getLanguage = function (): string {
    let nav = window.navigator, i, language: string;

    if (Array.isArray(nav.languages)) {
        for (i = 0; i < nav.languages.length; i++) {
            language = nav.languages[i];
            if (language && language.length) {
                return language.split("-")[0];
            }
        }
    }

    if (nav.language && nav.language.length) {
        return nav.language.split("-")[0];
    }

    return 'en'; //Default language is english
};

export function formatString(text: string, args: {[id: string]: any} = {}): string {
    return text.replace(/{(\d+)}/g, (match, number) => {
        return typeof args[number] != 'undefined'? args[number]: match;
    });
}

const I18nHelper = {
    getLanguage,
    formatString
}

export default I18nHelper;