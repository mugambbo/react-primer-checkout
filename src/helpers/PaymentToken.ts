export interface PaymentClientToken {
    accessToken: string;
    intent: string;
    env: string;
}

export interface PaymentAPIToken {
    clientToken: string;
    expirationDate: string;
}

export interface PaymentTokenizationResponse {
    deleted: boolean; //true or false
    createdAt: string; //"2021-05-04T09:33:52.293855",
    updatedAt: string; //"2021-05-04T09:33:52.293855",
    deletedAt: null;
    token: string; //"alooYT5lS4m2JdL60rNo9nwxNjIwMTIwODMy",
    analyticsId: string; //"-NxZbWYcXPiRMynOCdzu2G5L",
    tokenType: string; //"SINGLE_USE",
    paymentInstrumentType: string; //"PAYMENT_CARD",
    paymentInstrumentData:{
       last4Digits: string; //"1111",
       expirationMonth: string; //"03",
       expirationYear: string; //"2030",
       cardholderName: string; //"J Doe",
       network: string; //"Visa",
       isNetworkTokenized: boolean; //false,
       binData:{
          network: string; //"VISA",
          issuerCountryCode: string; //"US",
          issuerName: string; //"JPMORGAN CHASE BANK, N.A.",
          issuerCurrencyCode: string; //null,
          regionalRestriction: string; //"UNKNOWN",
          accountNumberType: string; //"UNKNOWN",
          accountFundingType: string; //"UNKNOWN",
          prepaidReloadableIndicator: string; //"NOT_APPLICABLE",
          productUsageType: string; //"UNKNOWN",
          productCode: string; //"VISA",
          productName: string; //"VISA"
       }
    },
    vaultData: any; //null,
    threeDSecureAuthentication: any; //null    
}