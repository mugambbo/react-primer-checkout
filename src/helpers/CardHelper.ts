import { Card, CardTypes } from './CardTypes';

/**
 * Strip all input characters that are not within the range of 0-9
 * @param string The alphanumeric 
 * @returns {string}
 */
export function numbersOnlyString (string: string): string {
  var numbersOnlyString = "";
  for(var i = 0; i < string.length; i++) {
    var currentChar = string.charAt(i);
    var isValid = !isNaN(parseInt(currentChar));
    if(isValid) { numbersOnlyString += currentChar; }
  }
  return numbersOnlyString;
};


/**
 * Apply a format mask to the string
 * @param string The string to be masked
 * @param mask The mask to be used
 * @returns {string} The string with the mask applied
 */
export function applyFormatMask (string: string, mask?: string): string {
  var formattedString = "";
  var numberPos = 0;
  if (!mask) return string;
  for(var j = 0; j < mask.length; j++) {
    var currentMaskChar = mask[j];
    if(currentMaskChar === "X") {
      var digit = string.charAt(numberPos);
      if(!digit) {
        break;
      }
      formattedString += string.charAt(numberPos);
      numberPos++;
    } else {
      formattedString += currentMaskChar;
    }
  }
  return formattedString;
};


/**
 * Establish the type of a card from the number.
 * @param number The card number
 * @returns {Card|null} The card type as string or undefined if not exist
 */
export function whatCardType (number: string): Card | undefined {
  let re: RegExp;
  // Diners
  re = new RegExp(CardTypes.DINERS.pattern);
  if (number.match(re) != null)
    return CardTypes.DINERS;

  // JCB
  re = new RegExp(CardTypes.JCB.pattern);
  if (number.match(re) != null)
    return CardTypes.JCB;

  // AMEX
  re = new RegExp(CardTypes.AMEX.pattern);
  if (number.match(re) != null)
    return CardTypes.AMEX;

  // Visa
  re = new RegExp(CardTypes.VISA.pattern);
  if (number.match(re) != null)
    return CardTypes.VISA;

  // Mastercard
  re = new RegExp(CardTypes.MASTERCARD.pattern);
  if (number.match(re) != null)
    return CardTypes.MASTERCARD;

  // Discover
  re = new RegExp(CardTypes.DISCOVER.pattern);
  if (number.match(re) != null)
    return CardTypes.DISCOVER;

  //Hipercard
  re = new RegExp(CardTypes.HIPERCARD.pattern);
  if (number.match(re) != null)
    return CardTypes.HIPERCARD;

  //UnionPay
  re = new RegExp(CardTypes.UNIONPAY.pattern);
  if (number.match(re) != null)
    return CardTypes.UNIONPAY;

  //Maestro Card
  re = new RegExp(CardTypes.MAESTROCARD.pattern);
  if (number.match(re) != null)
    return CardTypes.MAESTROCARD;

  //Verve Card
  re = new RegExp(CardTypes.VERVECARD.pattern);
  if (number.match(re) != null)
    return CardTypes.VERVECARD;

  return undefined;
};

/**
 * Checks whether the given card expiry (month and year) is valid?
 * @param month The expiry month on the card
 * @param year The expiry year on the card
 * @returns {boolean} True if expiry is valid, false otherwise
 */
 export function isExpiryValid(month: string, year: string): boolean {
  const today = new Date();
  const currentMonth = today.getMonth() + 1;
  const currentYear = String(today.getFullYear());
  const yearLong = currentYear.substring(0, 2) + year;

  const currentYearNum = parseInt(currentYear);
  const monthNum = parseInt(month);
  const yearNum = parseInt(yearLong);

  return (isValidMonth(monthNum) && (yearNum > currentYearNum)) || (yearNum === currentYearNum && monthNum >= currentMonth);
};


/**
 * Checks whether the given month is valid?
 * @param expiryMonth The expiry month on the card
 * @returns {boolean} True if expiry is valid, false otherwise
 */
 function isValidMonth (expiryMonth: number): boolean {
  return (expiryMonth >= 1 && expiryMonth <= 12);
};

/**
 * Checks whether the CVC of the card is valid
 * @return True if the cvc is valid, false otherwise
 */
export function isCVCValid (cvc: string, cardType?: Card): boolean {
    if (!cvc || !cardType) return false;
    const cvcValue = cvc.trim();
    return (cvcValue.length >= cardType.cvcMask.length) && Number(cvcValue) > 0;
}

export function buildPaymentInstrument(cardNumber: string, cardCVC: string, cardExpiry: string, cardName: string){
  return {
    paymentInstrument: {
      number: numbersOnlyString(cardNumber),
      cvv: cardCVC,
      expirationMonth: cardExpiry.split("/")[0].trim(),
      expirationYear: String(new Date().getFullYear()).substring(0, 2) + cardExpiry.split("/")[1].trim(),
      cardholderName: cardName
    }
  }
}