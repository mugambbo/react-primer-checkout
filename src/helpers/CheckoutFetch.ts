import Config from "../config";
import { PaymentClientToken } from "./PaymentToken";

async function post(path: string, clientToken: string, data = {}): Promise<any> {
    const url = `${Config.SDK_BASE_URL}${path}`;
    const accessTokenRes: PaymentClientToken = parseJwt(clientToken) as PaymentClientToken;

    const response = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Primer-Client-Token': accessTokenRes.accessToken
        },
        body: JSON.stringify(data)
    });

    // check for error response
    if (!response.ok) {
        const dataRes: {[key: string]: any}  = response.json();
        //Get error message from body or default to response statusText
        const error = (dataRes && dataRes.message) || response.statusText;
        throw new Error(error);
    }

    return response.json();
}

export function parseJwt(token: string): object {
    var payload64 = token.split('.')[1];
    var payload = Buffer.from(payload64, 'base64');
    return JSON.parse(payload.toString());
}

const PrimerFetch = {
    post
}

export default PrimerFetch;