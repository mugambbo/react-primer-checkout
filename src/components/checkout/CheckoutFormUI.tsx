import React, { ReactNode } from 'react';
import { Card } from '../../helpers/CardTypes';
import { PayButtonStyles } from './CheckoutContainer';
import PayButton from '../pay/PayButton';
import CardField from '../card';
import CardName from '../holder';

interface CheckoutFormUIProps {
  readonly btnStyles?: PayButtonStyles;
  readonly amount?: number;
  readonly currency?: string;
  readonly style?: object;
  readonly policy?: ReactNode;
  readonly cardNumber: string;
  readonly cardExpiry: string;
  readonly cardCVC: string;
  readonly cardName: string;
  readonly cardErrors: object;
  readonly inputStyles?: object;
  readonly cardNameError: string;
  readonly cardType: Card;
  readonly loading: boolean;
  onCardNameChange(event: React.ChangeEvent): void;
  onCardChange(event: React.ChangeEvent): void
  onPay?(event: React.FormEvent<HTMLFormElement>): void;
}

function CheckoutFormUI({btnStyles, amount, currency, style, policy: Policy, cardType, cardName, cardNumber, cardExpiry, cardCVC, cardNameError, cardErrors, inputStyles, loading, onCardNameChange, onCardChange, onPay}: CheckoutFormUIProps) {
  const primaryError = Object.values(cardErrors).filter((val) => !!val)[0]?? cardNameError;

  return (
    <form className="full-width" style={{minWidth: '400px', ...style }} onSubmit={onPay} >
      <div className="form-subcontainer" >
        <CardName loading={loading} cardName={cardName} inputStyles={{...inputStyles}} onCardNameChange={onCardNameChange} />
        <CardField loading={loading} cardType={cardType} cardNumber={cardNumber} cardExpiry={cardExpiry} cardCVC={cardCVC} inputStyles={inputStyles} onCardChange={onCardChange} />
        {Policy}
        <PayButton loading={loading} btnStyles={btnStyles} amount={amount} currency={currency} />
        <span className={primaryError? "helper-error": "helper"}>{primaryError}</span>
      </div>
    </form>
  )
}

export default CheckoutFormUI;