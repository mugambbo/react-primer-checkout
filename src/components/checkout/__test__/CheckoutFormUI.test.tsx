import React from 'react';
import { shallow } from 'enzyme';
import CheckoutFormUI from '../CheckoutFormUI';
import { CardTypes } from '../../../helpers/CardTypes';

describe('<CheckoutFormUI />', () => {
    const onCardChange = jest.fn();
    const onCardNameChange = jest.fn();

    test('renders the checkout ui', () => {    
        const component = shallow(<CheckoutFormUI cardType={CardTypes.GENERIC} cardNumber="" cardExpiry="" cardCVC="" loading={false} cardName="" cardErrors={{}} cardNameError="" onCardNameChange={onCardNameChange} onCardChange={onCardChange} />);  
        expect(component).toMatchSnapshot();
    });
});