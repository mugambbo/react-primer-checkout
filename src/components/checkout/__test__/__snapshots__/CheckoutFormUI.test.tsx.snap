// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutFormUI /> renders the checkout ui 1`] = `
<form
  className="full-width"
  style={
    Object {
      "minWidth": "400px",
    }
  }
>
  <div
    className="form-subcontainer"
  >
    <CardName
      cardName=""
      inputStyles={Object {}}
      loading={false}
      onCardNameChange={[MockFunction]}
    />
    <CardField
      cardCVC=""
      cardExpiry=""
      cardNumber=""
      cardType={
        Object {
          "cvcMask": "XXX",
          "mask": "XXXX XXXX XXXX XXXX",
          "name": "GENERIC",
          "pattern": "",
        }
      }
      loading={false}
      onCardChange={[MockFunction]}
    />
    <PayButton
      loading={false}
    />
    <span
      className="helper-error"
      id="helper-container"
    />
  </div>
</form>
`;
